[FLINK-20456][docs] Make streaming SQL concepts more approachable

sjwiesman opened the pull request:
## What is the purpose of the change

The streaming concepts section of the documentation is very negative and focuses on the difficulties of SQL. While we don't want to lie to users about edge cases, or things that may be tricky, I think we can make the experience of getting started more approachable. 


## Brief change log

See commit messages. 


## Verifying this change


This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

No

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

sjwiesman commented:
@twalthr @wuchong @morsapaes @alpinegizmo @knaufk 

I think all of you would have good opinions on these changes.

alpinegizmo commented:
This needs a much more careful review than I've done so far, but I'll share a few initial impressions:

I prefer the original explanation of Relational Queries on Data Streams in the Dynamic Tables page. I find it easier to understand. 

On the Time Attributes page I like putting event time before processing time. However, I am wondering about the deletion of the TableSource sub-sections. I don't like having those sections here -- these are low-level details that get in the way of understanding -- but is this valuable information that is being lost?

I mostly like what's going on in the page on joins, but at first I was left confused by the relationship between versioned tables and temporal joins. I had to read pretty closely to figure out what's going on; I think there are too many similar concepts with related names for this to be easy to grasp. I think we need to think carefully about how to organize the docs so that the relationships are clear. 

I suggest we take the section on Temporal Table Functions out of the page on versioned tables, along with the section on joins with temporal table functions in the page on Joins in Continuous Queries, and put those two sections into their own "THIS PAGE DESCRIBES A LEGACY FEATURE" page by themselves. 

We also need to do more than just move things around. Consider this note concerning the phony temporal table function example `SELECT * FROM Rates('11:00:00');`:

> Note: Currently, Flink doesn’t support directly querying the temporal table functions with a constant time attribute parameter. The above example was used to provide an intuition about what the function Rates(timeAttribute) returns.

This old note is nonsense, since temporal table functions are a legacy feature we intend to phase out. But the tone of this note suggests that there might be further development.

FWIW, these pages have typos and grammar issues, most of which pre-date this PR.

sjwiesman commented:
Thanks for taking a look @alpinegizmo 

I went back and made a grammar pass on each page (except joins I haven't gotten that far yet). 

> I prefer the original explanation of Relational Queries on Data Streams in the Dynamic Tables page. I find it easier to understand.

I put the table back and just slightly tweaked the language to try and be more approachable

> On the Time Attributes page I like putting event time before processing time. However, I am wondering about the deletion of the TableSource sub-sections. I don't like having those sections here -- these are low-level details that get in the way of understanding -- but is this valuable information that is being lost?

The `TableSource` interface is deprecated, that's why I dropped it. 

I also completely rewrote the versioned table description. I think it's much better now.

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
-- *Processing time* refers to the system time of the machine (also known as "wall-clock time") that is executing the respective operation.
-- *Event time* refers to the processing of streaming data based on timestamps which are attached to each row. The timestamps can encode when an event happened.
-- *Ingestion time* is the time that events enter Flink; internally, it is treated similarly to event time.
+*Processing time* refers to the machine's system time (also known as "wall-clock time") that is executing the respective operation.
```
```suggestion
- *Processing time* refers to the machine's system time (also known as "wall-clock time") that is executing the respective operation.
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
+Time attributes can be part of every table schema.
+They are defined when creating a table from a `CREATE TABLE DDL` or a `DataStream`. 
+Once a time attribute is defined, it can be referenced as a field and used in time-based operations.
+As long as a time attribute is not modified and forwarded from one part of the query to another, it remains valid.
```
```suggestion
As long as a time attribute is not modified, and is simply forwarded from one part of a query to another, it remains a valid time attribute. 
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
+They are defined when creating a table from a `CREATE TABLE DDL` or a `DataStream`. 
+Once a time attribute is defined, it can be referenced as a field and used in time-based operations.
+As long as a time attribute is not modified and forwarded from one part of the query to another, it remains valid.
+Time attributes behave like regular timestamps and accessible for calculations.
```
```suggestion
Time attributes behave like regular timestamps, and are accessible for calculations.
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
+Once a time attribute is defined, it can be referenced as a field and used in time-based operations.
+As long as a time attribute is not modified and forwarded from one part of the query to another, it remains valid.
+Time attributes behave like regular timestamps and accessible for calculations.
+When used in computation, time attributes are materialized and act as standard timestamps. 
```
```suggestion
When used in calculations, time attributes are materialized and act as standard timestamps. 
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
+As long as a time attribute is not modified and forwarded from one part of the query to another, it remains valid.
+Time attributes behave like regular timestamps and accessible for calculations.
+When used in computation, time attributes are materialized and act as standard timestamps. 
+However, the opposite is not true; ordinary timestamp columns cannot be arbitrarily converted to time attributes in the middle of a query.
```
```suggestion
However, ordinary timestamps cannot be used in place of, or be converted to, time attributes.
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
 ----------
 
-Event time allows a table program to produce results based on the time that is contained in every record. This allows for consistent results even in case of out-of-order events or late events. It also ensures replayable results of the table program when reading records from persistent storage.
+Event time allows a table program to produce results based on timestamps in every record, allowing for consistent results even in case of out-of-order events or late events. It also ensures the replayability of the results of the table program when reading records from persistent storage.
```
Can we really argue that results are consistent with late events?

```suggestion
Event time allows a table program to produce results based on timestamps in every record, allowing for consistent results despite out-of-order or late events. It also ensures the replayability of the results of the table program when reading records from persistent storage.
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -22,213 +22,42 @@ specific language governing permissions and limitations
+Additionally, event time allows for unified syntax for table programs in both batch and streaming environments. A time attribute in a streaming environment can be a regular column of a row in a batch environment.
 
-In order to handle out-of-order events and distinguish between on-time and late events in streaming, Flink needs to extract timestamps from events and make some kind of progress in time (so-called [watermarks]({% link dev/event_time.md %})).
+To handle out-of-order events and distinguish between on-time and late events in streaming, Flink needs to extract timestamps from events and make some progress in time (so-called [watermarks]({% link dev/event_time.md %})).
```
```suggestion
To handle out-of-order events and to distinguish between on-time and late events in streaming, Flink needs to know the timestamp for each row, and it also needs regular indications of how far along in event time the processing has progressed so far (via so-called [watermarks]({% link dev/event_time.md %})).
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -251,14 +80,12 @@ GROUP BY TUMBLE(user_action_time, INTERVAL '10' MINUTE);
 ### During DataStream-to-Table Conversion
 
-The event time attribute is defined with the `.rowtime` property during schema definition. [Timestamps and watermarks]({% link dev/event_time.md %}) must have been assigned in the `DataStream` that is converted.
+When converting a `DataStream` to a table, an event time attribute can be defined with the `.rowtime` property during schema definition. [Timestamps and watermarks]({% link dev/event_time.md %}) must have been already assigned in the `DataStream` being converted.
```
```suggestion
When converting a `DataStream` to a table, an event time attribute can be defined with the `.rowtime` property during schema definition. [Timestamps and watermarks]({% link dev/event_time.md %}) must have already been assigned in the `DataStream` being converted.
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -251,14 +80,12 @@ GROUP BY TUMBLE(user_action_time, INTERVAL '10' MINUTE);
+When converting a `DataStream` to a table, an event time attribute can be defined with the `.rowtime` property during schema definition. [Timestamps and watermarks]({% link dev/event_time.md %}) must have been already assigned in the `DataStream` being converted.
 
-There are two ways of defining the time attribute when converting a `DataStream` into a `Table`. Depending on whether the specified `.rowtime` field name exists in the schema of the `DataStream` or not, the timestamp field is either
+There are two ways of defining the time attribute when converting a `DataStream` into a `Table`. Depending on whether the specified `.rowtime` field name exists in the schema of the `DataStream`, the timestamp is either (1) appended as a new column to the schema or
+(2) replaces an existing column.
```
```suggestion
There are two ways of defining the time attribute when converting a `DataStream` into a `Table`. Depending on whether the specified `.rowtime` field name exists in the schema of the `DataStream`, the timestamp is either (1) appended as a new column, or it
(2) replaces an existing column.
```

alpinegizmo commented on docs/dev/table/streaming/time_attributes.md:
```diff
@@ -318,98 +145,61 @@ val windowedTable = table.window(Tumble over 10.minutes on $"user_action_time" a
+Processing Time
+---------------
+
+Processing time allows a table program to produce results based on the time of the local machine. It is the simplest notion of time but does not provide determinism. It neither requires timestamp extraction nor watermark generation.
```
```suggestion
Processing time allows a table program to produce results based on the time of the local machine. It is the simplest notion of time, but it will generate non-deterministic results. Processing time does not require timestamp extraction or watermark generation.
```

alpinegizmo reviewed (commented):
Some specific suggestions; hoping to make these pages easier to understand.

sjwiesman commented:
thank you @alpinegizmo I've applied your suggestions inline.

sjwiesman commented:
If there aren't any other comments, I'll merge this tomorrow.

alpinegizmo reviewed (changes requested):
This isn't ready to merge yet. I've pointed out some specific problems on the page about versions tables. These have easy fixes, which I've supplied as suggestions where possible.

There's one more thing: the Legacy section in the page on Joins talks a bit about joins with Temporal Table Functions, but it's largely incomprehensible. In this context, I'm concerned about dropping the docs on Temporal Table Functions, which this PR deletes from the page on versioned tables. I'm happy to see them leave that page, but I'm not sure we can simply drop them.

sjwiesman commented on docs/dev/table/streaming/temporal_tables.md:
```diff
@@ -57,134 +56,68 @@ SELECT * FROM product_changelog;
-To define a versioned table on `RatesHistory`, Flink supports defining a versioned view 
-by [deduplication query]({% link dev/table/sql/queries.md %}#deduplication) which produces an ordered changelog
-stream with an inferred primary key(`currency`) and event time(`currency_time`).
+Flink can reinterpret this table as a versioned table by defining a [deduplication query]({% link dev/table/sql/queries.md %}#deduplication) which produces an ordered changelog stream with an inferred primary key (currency) and event time (update_time). 
```
The real answer is flink has a special optimization step that turns any `OVER WINDOW` with exactly this format into a versioned table keyed by the fields in the `PARTITION BY` clause. 

```sql
SELECT *,
      ROW_NUMBER() OVER (PARTITION BY [key,] ORDER BY [time attribute] DESC) AS row_num
WHERE 
     row_num = 1
```

In a batch query, this would return the most recent value for each currency. The streaming semantic is it provides the new value for each partition each time a more recent value is made available.

sjwiesman commented:
@alpinegizmo thank you for your patience 

I've updated the versioned view section. 

![Screen Shot 2020-12-08 at 10 54 02 AM](https://user-images.githubusercontent.com/1891970/101515251-c4116b00-3943-11eb-9732-c93e381b8532.png)

I also moved temporal table functions to a new "legacy" page and rewrote the example. 

![Screen Shot 2020-12-08 at 11 44 38 AM](https://user-images.githubusercontent.com/1891970/101520936-c3300780-394a-11eb-82ce-6d20a88cdacf.png)

alpinegizmo commented on docs/dev/table/streaming/legacy.md:
```diff
@@ -0,0 +1,135 @@
+
+The temporal table function is the legacy way of defining something akin to a [versioned table]({% link dev/table/streaming/temporal_tables.md %})
+that can be used in a temporal table join.
+Please define temporal joins using [versioned table]({% link dev/table/streaming/temporal_tables.md %}) as described above in new queries.
```
Temporal tables are no longer "above".

```suggestion
Please define temporal joins using [versioned tables]({% link dev/table/streaming/temporal_tables.md %}) in new queries.
```

alpinegizmo commented on docs/dev/table/streaming/legacy.md:
```diff
@@ -0,0 +1,135 @@
+
+#### Defining a Temporal Table Function
+
+Temporal table function can be defined on top of append-only streams using the [Table API]({% link dev/table/tableApi.md %}).
```
```suggestion
Temporal table functions can be defined on top of append-only streams using the [Table API]({% link dev/table/tableApi.md %}).
```

alpinegizmo commented on docs/dev/table/streaming/legacy.md:
```diff
@@ -0,0 +1,135 @@
+Append-only tables (left input/probe side) can join with a temporal table (right input/build side),
+i.e., a table that changes over time and tracks its changes, to retrieve the value for a key as it was at a particular point in time.
+
+Consider an append-only table `orders` that tracks customers orders in different currencies.
```
```suggestion
Consider an append-only table `orders` that tracks customers' orders in different currencies.
```

alpinegizmo reviewed (commented):
In general this is looking good.

Spotted a couple of small things.

Also, I think it might be a preferable to rename temporal_tables.md to versioned_tables.md. It's unfortunate that the page title no longer matches the filename. But doing this would mean checking and updating any links to this page (and maybe even adding a redirect).

sjwiesman commented:
> Also, I think it might be a preferable to rename temporal_tables.md to versioned_tables.md. It's unfortunate that the page title no longer matches the filename. But doing this would mean checking and updating any links to this page (and maybe even adding a redirect).

Yeah, that's why I avoided it 🙈  I'll make the change

sjwiesman commented:
@alpinegizmo I've applied your changes, rebased, and squashed the fixup. Let me know if you have any other concerns.

sjwiesman commented:
I also renamed the file and added a redirect.

alpinegizmo reviewed (approved):
I'll approve this now, but there remains one issue to deal with (regarding the redirect).

sjwiesman commented on docs/redirects/temporal_table.md:
```diff
@@ -0,0 +1,24 @@
+---
+title: Temporal Tables
+layout: redirect
+redirect: /dev/table/streaming/versioned_table.html
```
damn, thank you

sjwiesman replied:
- Now that you say it, I think legacy makes more sense. That's the content people are expecting to find. And that page already cross-links to versioned_tables. 
- The `zh` redirect happens automatically

sjwiesman commented:
thanks for the review, will merge.
